src/t8_data/t8_element_array_iterator.hxx

```diff
@@ -0,0 +1,190 @@
+/*
+  This file is part of t8code.
+  t8code is a C library to manage a collection (a forest) of multiple
+  connected adaptive space-trees of general element classes in parallel.
+
+  Copyright (C) 2015 the developers
+
+  t8code is free software; you can redistribute it and/or modify
+  it under the terms of the GNU General Public License as published by
+  the Free Software Foundation; either version 2 of the License, or
+  (at your option) any later version.
+
+  t8code is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with t8code; if not, write to the Free Software Foundation, Inc.,
+  51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
+*/
+
+/** \file t8_element_array_iterator.hxx
+ * We define a bidirectional-iterator for a \a t8_element_array_t that stores pointers to elements of a given
+ * eclass scheme.
+ */
+
+#ifndef T8_ELEMENT_ARRAY_ITERATOR_HXX
+#define T8_ELEMENT_ARRAY_ITERATOR_HXX
+
+#include <t8.h>
+#include <t8_element.h>
+#include <t8_data/t8_containers.h>
+
+#include <cstddef>
+#include <iterator>
+
+T8_EXTERN_C_BEGIN ();
+
+/**
+ * \brief This class resembles a bidirectional-iterator wrapper for the content of a \a t8_element_array_t.
+ * The iterators can be dereferenced in order to receive an \a t8_element_t pointer to the corresponding index.
+ * These iterators may be used in algorithms of the standard library (e.g. std::lower_bound,...) or to iterate
+ * through an \a t8_element_array_t.
+ * Since the actual data is serialized to char-bytes in the underlying array, we reconstruct a \a t8_element_t pointer
+ * and let the dereference operator return it as a value_type. Therefore, read-only operations on the
+ * \a t8_element_array_t are possible.
+ */
+class t8_element_array_iterator {
+
+ private:
+  const t8_eclass_scheme_c* scheme; /*!< The scheme of the elements residing within the array. */
+  const sc_array_t* elements;       /*!< A pointer to the actual serialized array of element pointers. */
+  t8_locidx_t current_index { 0 };  /*!< The index the iterator currently points to. */
+
+ public:
+  using iterator_category = std::bidirectional_iterator_tag;
```

Comment:
For proper binary search support you need to change to this.
```suggestion
  using iterator_category = std::random_access_iterator_tag;
```

https://en.cppreference.com/w/cpp/algorithm/binary_search states that
```
However, if ForwardIt is not a [LegacyRandomAccessIterator](https://en.cppreference.com/w/cpp/named_req/RandomAccessIterator), the number of iterator increments is linear in N.
```

Reply:
Then more operators are needed.

Reply:
Yes some of the current operators are needed in order to meet the requirements of a LegacyRandomAccessIterator, but apparently I have not implemented all of the needed comparison operators. However, this does not enforce us to switch to a std::random_access_iterator_tag in order to not have a linear complexity in the iterator increments. However, we can try to change the iterator from a bidirectional to random access iterator nonetheless.

Reply:
I suggest to do so. Otherwise we would be slower than the old code, right?

Reply:
Yes, this could be the case. By switching to a std::random_iterator_tag we can be sure that it has a logarithmic complexity in the iterator increments.